src/Cardano/AddressDerivation/Byron.hs

```diff
@@ -0,0 +1,252 @@
+
+      -- * Derivation
+    , deriveAccountPrivateKey
+    , deriveAddressPrivateKey
+
+    ) where
+
+import Prelude
+
+import Cardano.AddressDerivation
+    ( Depth (..), DerivationType (..), Index (..) )
+import Cardano.Crypto.Wallet
+    ( DerivationScheme (DerivationScheme1)
+    , XPrv
+    , XPub
+    , deriveXPrv
+    , generate
+    , toXPub
+    , unXPub
+    )
+import Cardano.Mnemonic
+    ( SomeMnemonic (..), entropyToBytes, mnemonicToEntropy )
+import Control.DeepSeq
+    ( NFData )
+import Crypto.Hash
+    ( hash )
+import Crypto.Hash.Algorithms
+    ( Blake2b_256, SHA512 (..) )
+import Data.ByteArray
+    ( ScrubbedBytes )
+import Data.ByteString
+    ( ByteString )
+import GHC.Generics
+    ( Generic )
+
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Write as CBOR
+import qualified Crypto.KDF.PBKDF2 as PBKDF2
+import qualified Data.ByteArray as BA
+
+{-------------------------------------------------------------------------------
+                                   Key Types
+-------------------------------------------------------------------------------}
+
+-- | Material for deriving HD random scheme keys, which can be used for making
+-- addresses.
+data ByronKey (depth :: Depth) key = ByronKey
+    { getKey :: key
+    -- ^ The raw private or public key.
+    , derivationPath :: DerivationPath depth
+    -- ^ The address derivation indices for the level of this key.
+    , payloadPassphrase :: ScrubbedBytes
+    -- ^ Used for encryption of payload containing address derivation path.
+    } deriving stock (Generic)
+
+instance (NFData key, NFData (DerivationPath depth)) => NFData (ByronKey depth key)
+deriving instance (Show key, Show (DerivationPath depth)) => Show (ByronKey depth key)
+deriving instance (Eq key, Eq (DerivationPath depth)) => Eq (ByronKey depth key)
+
+-- | The hierarchical derivation indices for a given level/depth.
+type family DerivationPath (depth :: Depth) :: * where
+    -- The root key is generated from the seed.
+    DerivationPath 'RootK =
+        ()
+    -- The account key is generated from the root key and account index.
+    DerivationPath 'AccountK =
+        Index 'WholeDomain 'AccountK
+    -- The address key is generated from the account key and address index.
+    DerivationPath 'AddressK =
+        (Index 'WholeDomain 'AccountK, Index 'WholeDomain 'AddressK)
+
+{-------------------------------------------------------------------------------
+                                 Key generation
+-------------------------------------------------------------------------------}
+
+-- | The amount of entropy carried by a BIP-39 12-word mnemonic is 16 bytes.
+minSeedLengthBytes :: Int
+minSeedLengthBytes = 16
+
+-- | Generate a root key from a corresponding seed.
+-- The seed should be at least 16 bytes.
+generateKeyFromSeed
+    :: SomeMnemonic
+    -> ScrubbedBytes
+    -> ByronKey 'RootK XPrv
+generateKeyFromSeed = unsafeGenerateKeyFromSeed ()
```

Comment:
This should really implement: https://input-output-hk.github.io/cardano-addresses/haddock/Cardano-AddressDerivation.html#t:GenMasterKey ?

Reply:
https://github.com/input-output-hk/cardano-addresses/pull/4/commits/22ce22c2de17a43e34d027ec0eb18933e086d814